docs/trailing-commas-in-calls.md

```diff
@@ -0,0 +1,121 @@
+# Trailing comma support in calls and function definitions
+
+## Summary
+
+Luau will support trailing commas in the following places:
+
+```lua
+local function definition(
+    x: number, -- <-- Here, previously a syntax error
+)
+
+call(
+    a,
+    b, -- <-- Here, previously a syntax error
+)
+
+type Fn = (
+    x: number,
+) -> ()
+```
+
+## Motivation
+
+Currently, trailing commas are supported in several places in Luau.
+
+```lua
+local t = {
+    a,
+    b,
+    c,
+}
+
+-- Which even means...
+f {
+    a,
+    b,
+    c,
+}
+
+type Type = {
+    a: number,
+    b: number,
+}
+```
+
+However, they are not supported in function definitions, types, or calls. This is an unintuitive syntax error and it is obvious what the user intended to do.
```

Comment:
> This is an unintuitive syntax error and it is obvious what the user intended to do.

I completely disagree, the way they are disallowed in function arguments is completely intuitive to me.
I would even say that I don't know what user might have intended if they left a trailing ',', as in I agree with the current compiler that they likely forgot to type the last argument.

The motivation for this feature is lacking and weak.

Reply:
There's two use cases.

With formatting, if the line containing the function name + arguments exceeds the line length, reformat into its multiline variant that's still visually pleasing on the eyes.

```lua
function long_function_name(with, really, long, arguments, some: or_all, might: have_types)
end
```
into:
```
function long_function_name(
  with,
  really,
  long,
  arguments,
  some: or_all,
  might: have_types,
)
end
```

With `git diff`, if you add one more argument, then you will only see that line changed, not the line prior to that argument. Say, if you had:

```lua
  function foo(
    x,
    y
  )
```
then if you add a new argument `z`:
```diff
  function foo(
    x,
+   y,
+   z,
  )
```
as opposed to:
```lua
  function foo(
    x,
    y,
  )
```
then if you add a new argument `z`:
```diff
  function foo(
    x,
    y,
+   z,
  )
```

Reply:
Yeah, personally I'm the sort of person who would require it in a style guide when arguments are split across lines (since it makes reordering lines easier, in addition to the git diff problem). So, I'm personally sympathetic to supporting it. I wouldn't be too upset if we rejected it in single-line definitions too, but it's easier to just always allow the trailing comma obviously.

Reply:
I feel trailing commas should be allowed in both single & multiline calls, for consistency reasons. Luau isn't a whitespace-delimited language, so if you're getting a syntax error that disappears the moment you press Enter, it just seems kinda arbitrary. If tables can get an extra comma without needing a newline, why not function calls?

Reply:
I don't think consistency argument holds here, simply because function calls and tables are different language constructs with different uses.

Reply:
Tables and function calls are pretty different things—agreed, but since this is a syntax thing, I was looking at it from a broader syntax angle: 

In this case I interpret consistency as visual consistency, 

Using the duck method, a function call and a table definition are both visually comma-separated-lists, placed within some form of bracket (round or curly) delimiters. A user might want to use this fact to organize inputs with linebreaks; a contrived example: 
```lua
local o1, o2, o3, o4, o5, o6, o7 = getSomeClassObjectsAsTuple()
local composite123, composite456 = getCompositesFromTriples(
    o1, o2, o3,
    o4, o5, o6, -- arguments
    -- space to put o7 → o9 to make 3x3 matrix of args later
)
-- a similarly formatted table array definition:

local xTriples = {
    o1, o4, o7,
}
```
I feel that from a visual syntax perspective, both look like comma separated lists before they look like tables or function calls specifically, so ∴ IMHO it makes sense both forms of list adopt the same trailing-comma syntax.

Reply:
Also — just realized I might've been a bit vague the other day with the "arbitrary" 

I meant that in response to @aatxe that if we allow a multiline call to end with a `,`
```lua
call(
  arg1,
  arg2,
  arg3,
)
```
Then I believe we should also allow a single-line call to end with a `,`:
`call(arg1, arg2, arg3, )`

Otherwise, you'd get a whitespace-sensitve syntax error that disappears and reappears with a linebreak, and cases like: 
```lua
call(arg1, arg2, arg3,
) -- not cause SyntaxError?

call(arg1,
arg2, arg3,) -- SyntaxError?
```
the easy solution is just allow all of those cases since style guides/conventions should handle the rest 😅.

Reply:
> Luau isn't a whitespace-delimited language, so if you're getting a syntax error that disappears the moment you press Enter, it just seems kinda arbitrary.

I want to note that this is not true for function calls:
```
-- no error
local f = attachLog(add)(1, 4)

-- error
local g = attachLog
(add)(1, 4)
```

This is not to suggest that we need more cases like this, I would prefer we avoid them.